src/modules/battery_resistance_est/battery_resistance_est.cpp

```diff
@@ -0,0 +1,260 @@
+	WorkItem(MODULE_NAME, px4::wq_configurations::lp_default)
+{
+}
+
+bool InternalRes::init()
+{
+	if (!_battery_sub.registerCallback()) {
+		PX4_ERR("battery status callback registration failed!");
+		return false;
+	}
+
+	return true;
+}
+
+float InternalRes::extract_parameters() {
+
+	float r_steady_state = param_0;
+	float r_transient = (param_1/param_2) - param_0;
+	float voltage_open_circuit = param_3/param_2;
+
+	//calculate bat1_r_internal
+	float internal_resistance = (voltage_filtered_v - voltage_open_circuit) / (-current_filtered_a);
+
+	//logging
+	inter_res.r_transient = r_transient;
+	inter_res.r_steady_state = r_steady_state;
+	inter_res.voltage_open_circuit = voltage_open_circuit;
+
+	return internal_resistance;
+
+}
+
+
+void InternalRes::write_internal_resistance(float voltage_estimation_error, float internal_resistance){
+
+
+	//store best esimate
+	if((abs(voltage_estimation_error)) <= abs(best_prediction_error))
+	{
+		best_prediction = bat1_r_internal_prev;
+		best_prediction_error = voltage_estimation_error;
+	} else if (best_prediction_error_reset){
+		best_prediction = bat1_r_internal_prev;
+		best_prediction_error = voltage_estimation_error;
+		best_prediction_error_reset = false;
+	}
+
+	float time_since_param_write = (hrt_absolute_time()- last_param_write_time)/ 1e6f;
+
+	// save bat1_r_internal only periodically
+	if (time_since_param_write >= _inter_res_update_period.get()){
+
+		//BAT${i}_R_INTERNAL  increment: 0.01
+		best_prediction = round(best_prediction*100)/100;
+
+		//clamp BAT${i}_R_INTERNAL min: -1.0  max: 0.2
+		if (best_prediction > 0.2f){
+	    		best_prediction = 0.2f;
+		}
+		if (best_prediction < -1.0f){
+	    		best_prediction = -1.0f;
+		}
+
+		//Set px4 Internal resistance using simplified Rint Model
+		_bat1_r_internal.set(best_prediction);
+		_bat1_r_internal.commit();
+		last_param_write_time = hrt_absolute_time();
+		best_prediction_error_reset = true;
+	}
+
+	bat1_r_internal_prev = internal_resistance;
+
+	//logging
+	inter_res.bat1_r_internal = internal_resistance;
+
+}
+
+void InternalRes::estimate_v_dot(float voltage_estimation_error, float battery_sampling_period){
+
+	//process signal
+	signal_0 = -(battery_status.current_filtered_a - current_filtered_a_prev)/((battery_status.timestamp - battery_time_prev)/ 1e6f); //central difference method
+	signal_1 = -current_filtered_a;
+	signal_2 = -voltage_estimation_prev;
+
+	//update parameter vector estimate
+	param_0 += adaptation_gain_0*voltage_estimation_error*signal_0*battery_sampling_period;
+	param_1 += adaptation_gain_1*voltage_estimation_error*signal_1*battery_sampling_period;
+	param_2 += adaptation_gain_2*voltage_estimation_error*signal_2*battery_sampling_period;
+	param_3 += adaptation_gain_3*voltage_estimation_error*battery_sampling_period;
+
+	//Apply thevenin battery model (first order system)
+	float v_dot_estimate = param_0*signal_0 + param_1*signal_1 + param_2*signal_2 + param_3 + lambda*voltage_estimation_error;
```

Comment:
For more clarity, you could use the Matrix library as follows:

```suggestion
void InternalRes::estimate_v_dot(float voltage_estimation_error, float dt){

	//process signal
	Vector<float, 4> signal;
	signal(0) = -(battery_status.current_filtered_a - current_filtered_a_prev) / ((battery_status.timestamp - battery_time_prev) / 1e6f); //central difference method
	signal(1) = -current_filtered_a;
	signal(2) = -voltage_estimation_prev;
	signal(3) = 1.f;

	//update parameter vector estimate
	for (int i = 0; i < 4; i++) {
		_param_est(i) += _adaptation_gain(i) * voltage_estimation_error * signal(i) * dt;
	}

	//Apply thevenin battery model (first order system)
	const float v_dot_estimate = _param_est.transposed() * signal + _lambda * voltage_estimation_error;
```

This will significantly reduce the required number of variables and improve readability.

Reply:
@bresch great idea! I made the suggested changes. Thanks!